src/sentry/grouping/fingerprinting/__init__.py

```diff
@@ -470,17 +381,112 @@ def text(self):
+                yield from base_rules
+
+    def get_fingerprint_values_for_event(self, event: dict[str, Any]) -> None | Any:
+        if not (self.bases or self.rules):
+            return
+        access = EventAccess(event)
+        for rule in self.iter_rules():
+            new_values = rule.get_fingerprint_values_for_event_access(access)
+            if new_values is not None:
+                return (rule,) + new_values
+        return None
+
+    @classmethod
+    def _from_config_structure(cls: Any, data: dict[str, Any], bases: Any = None) -> Any:
+        version = data["version"]
+        if version != VERSION:
+            raise ValueError("Unknown version")
+        return cls(
+            rules=[Rule._from_config_structure(x) for x in data["rules"]],
+            version=version,
+            bases=bases,
+        )
+
+    def _to_config_structure(self, include_builtin: bool = False) -> None:
+        rules = self.iter_rules(include_builtin=include_builtin)
+
+        return {"version": self.version, "rules": [x._to_config_structure() for x in rules]}
+
+    def to_json(self, include_builtin: bool = False) -> None:
+        return self._to_config_structure(include_builtin=include_builtin)
+
+    @classmethod
+    def from_json(cls, value: dict[str, Any], bases: Any = None) -> Any:
+        try:
+            return cls._from_config_structure(value, bases=bases)
+        except (LookupError, AttributeError, TypeError, ValueError) as e:
+            raise ValueError("invalid fingerprinting config: %s" % e)
+
+    @staticmethod
+    def from_config_string(s: Any, bases: Any = None) -> Any:
+        try:
+            tree = fingerprinting_grammar.parse(s)
+        except ParseError as e:
+            context = e.text[e.pos : e.pos + 33]
+            if len(context) == 33:
+                context = context[:-1] + "..."
+            raise InvalidFingerprintingConfig(
+                f'Invalid syntax near "{context}" (line {e.line()}, column {e.column()})'
+            )
+        return FingerprintingVisitor(bases=bases).visit(tree)
+
+
+class BuiltInFingerprintingRules(FingerprintingRules):
+    """
+    A FingerprintingRules object that marks all of its rules as built-in
+    """
+
+    @staticmethod
+    def from_config_string(s: Any, bases: Any = None) -> None:
+        fingerprinting_rules = FingerprintingRules.from_config_string(s, bases=bases)
+        for r in fingerprinting_rules.rules:
+            r.is_builtin = True
+        return fingerprinting_rules
+
+    @classmethod
+    def _from_config_structure(
+        cls: Any, data: dict[str, Any], bases: Any = None
+    ) -> Sequence[FingerprintingRules]:
+        fingerprinting_rules = super()._from_config_structure(data, bases=bases)
+        for r in fingerprinting_rules.rules:
+            r.is_builtin = True
+        return fingerprinting_rules
+
+
+class FingerprintingVisitor(NodeVisitor):  # type: ignore[type-arg]
```

Comment:
If I do `NodeVisitor[FingerprintingRules]` then the tests fail, thus, I'm using this ignore.

Reply:
you can do:

```python
if TYPE_CHECKING:
   NodeVisitorBase = NodeVisitor[FingerprintingRules]
else:
    NodeVisitorBase = NodeVisitor


class FingerprintingVisitor(NodeVisitorBase):
    ...
```

and then it should be fine with it